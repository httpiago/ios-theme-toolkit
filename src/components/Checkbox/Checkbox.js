import React from 'react'
import PropTypes from 'prop-types'
import { mapClassList } from '../../Utils'

/**
 * Checkbox component.
 *
 * @since 0.1.0
 * @see https://httpiago.github.io/ios-theme-toolkit/#/components/Checkbox
 */
export default function Checkbox({ color, size, className: aditionalClasses, ...rest }) {

  const prefix = 'checkbox'
  const classes = mapClassList({
    [`${prefix}--${color}`]: color,
    [`${prefix}--${size}`]: size,
  })

  return (
    <input
      type="checkbox"
      className={`${prefix} ${classes} ${aditionalClasses}`}
      style={(typeof size === 'number') ? { height: size, width: size } : {}}
      {...rest}
    />
  )
}

Checkbox.defaultProps = {
  className: '',
  color: 'default',
  size: 'default'
}

Checkbox.propTypes = {
  /** State of checkbox for CONTROLLED components. */
  checked: PropTypes.bool,
  /** The initial checkbox state for UNCONTROLLED components. */
  defaultChecked: PropTypes.bool,
  /** Aditional classes. */
  className: PropTypes.string,
  /** Color of checkbox. */
  color: PropTypes.oneOf([ 'default', 'blue', 'red', 'green', 'yellow', 'gray' ]),
  /** Size of checkbox. */
  size: PropTypes.oneOfType([ PropTypes.number, PropTypes.oneOf([ 'default', 'small', 'big' ]) ]),
  /** https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input#attr-name */
  name: PropTypes.string,
}
