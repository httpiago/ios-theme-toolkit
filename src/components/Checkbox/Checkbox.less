@import '../../customizations.less';

// Default styles
.checkbox {
  -webkit-appearance: none;
  appearance: none;
  background-color: #FFF;
  border: 1px solid #c7c7cc;
  border-radius: 24%;
  position: relative;
  height: 22px;
  width: 22px;
  cursor: pointer;
  outline: 0;
  vertical-align: middle;
  -webkit-box-sizing: border-box;
  box-sizing: border-box;
  vertical-align: middle;
}

// Checked state
.checkbox:checked {
  border: none;
  background-color: @blue-color;
  background-repeat: no-repeat;
  background-position: center;
  background-image: url(data:image/svg+xml;base64,PHN2ZyB4bWxucz0naHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmcnIHg9JzBweCcgeT0nMHB4JyB2aWV3Qm94PScwIDAgMTIgOScgeG1sOnNwYWNlPSdwcmVzZXJ2ZSc+PHBvbHlnb24gZmlsbD0nI2ZmZmZmZicgcG9pbnRzPScxMiwwLjcgMTEuMywwIDMuOSw3LjQgMC43LDQuMiAwLDQuOSAzLjksOC44IDMuOSw4LjggMy45LDguOCAnLz48L3N2Zz4=);
  background-size: 60%;
}

// focus
.checkbox:not(:active):focus {
  box-shadow: 0 0 0 3px fade(@blue-color, 30%);
}

.checkbox[disabled] {
  opacity: .6;
  pointer-events: none;
  cursor: default;
}

// Checkbox pequeno
.checkbox--small {
  height: 16px;
  width: 16px;
}

// Checkbox grande
.checkbox--big {
  height: 28px;
  width: 28px;
}

// Checkbox com cor azul
.checkbox--blue {
  &:checked {
    background-color: @blue-color;
  }

  &:not(:active):focus {
    box-shadow: 0 0 0 3px fade(@blue-color, 30%);
  }
}

// Checkbox com cor vermelha
.checkbox--red {
  &:checked {
    background-color: @red-color;
  }

  &:not(:active):focus {
    box-shadow: 0 0 0 3px fade(@red-color, 30%);
  }
}

// Checkbox com cor amarela
.checkbox--yellow {
  &:checked {
    background-color: @yellow-color;
  }

  &:not(:active):focus {
    box-shadow: 0 0 0 3px fade(@yellow-color, 30%);
  }
}

// Checkbox com cor verde
.checkbox--green {
  &:checked {
    background-color: @green-color;
  }

  &:not(:active):focus {
    box-shadow: 0 0 0 3px fade(@green-color, 30%);
  }
}

// Checkbox com cor cinza
.checkbox--gray {
  &:checked {
    background-color: @gray-color;
  }

  &:not(:active):focus {
    box-shadow: 0 0 0 3px fade(@gray-color, 30%);
  }
}

// Checkbox com cor branca
.checkbox--white {
  &:checked {
    background-color: @white-color;
  }

  &:not(:active):focus {
    box-shadow: 0 0 0 3px fade(@white-color, 30%);
  }
}
