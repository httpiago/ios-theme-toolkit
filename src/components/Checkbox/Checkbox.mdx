---
name: Checkbox
route: /components/Checkbox
menu: Componentes
---

import { Playground, PropsTable } from 'docz'
import { Checkbox } from '../../index.js'
import '../../../dist/styles.css'

# Checkbox

To use this component, simply import it into any file and pass the [modifiers](#/#props) in the properties.

```jsx
import { Checkbox } from 'ios-theme-toolkit'

export default () => (
  <Checkbox checked={true} />
)
```

## Examples

<Playground>
  <span className="desc">Default style and size (22px):</span>
  <Checkbox defaultChecked={true} />

  <span className="desc">Colorful:</span>
  <Checkbox defaultChecked={true} color="green" />
  <Checkbox defaultChecked={true} color="red" />
  <Checkbox defaultChecked={true} color="yellow" />
  <Checkbox defaultChecked={true} color="gray"/>
  <Checkbox defaultChecked={true} color="white"/>

  <span className="desc">Disabled:</span>
  <Checkbox defaultChecked={true} disabled />

  <span className="desc">Not checked:</span>
  <Checkbox defaultChecked={false} />

  <span className="desc">Small (16px):</span>
  <Checkbox defaultChecked={true} size="small"/>

  <span className="desc">Big (28px):</span>
  <Checkbox defaultChecked={true} size="big" />

  <span className="desc">Custom size (~px):</span>
  <Checkbox defaultChecked={true} size={100} />

  <style>{".react-live-preview .desc { display: block; font-weight: bold; margin-top: 20px; }"}</style>
  <style>{".react-live-preview .checkbox { margin: 0 3px; }"}</style>
</Playground>

## Props

<PropsTable of={ Checkbox } />

All the properties which are not listed above will be transferred to the <\input type="checkbox"\> tag, so, It accepts all props which native checkbox support.
