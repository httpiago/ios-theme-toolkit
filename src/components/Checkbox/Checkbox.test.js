/**
 * TEST FILE
 * Please try writing the tests for all component behaviors,
 * manually call the functions within the class, change the properties...
 *
 * @see https://jestjs.io/docs/en/expect
 * @see https://github.com/airbnb/enzyme/tree/master/docs/api
 */
import React from 'react'
import { shallow } from 'enzyme'
import Checkbox from './Checkbox'

const COMPONENT = shallow(
  <Checkbox className="aditionalClass" />
)

describe('Test Checkbox component', () => {
  it('should render correctly based on last snapshot', () => {
    expect(COMPONENT).toMatchSnapshot()
  })

  it('should update correctly', () => {
    COMPONENT.setProps({ color: 'green' })

    expect(COMPONENT.hasClass('checkbox--green')).toBeTruthy()
  })

  it('should has aditional classes', () => {
    expect(COMPONENT.hasClass('aditionalClass')).toBeTruthy()
  })

  it('should pass aditional props to html element', () => {
    COMPONENT.setProps({ 'data-custom-attr': 'yes' })

    expect(COMPONENT.html()).toContain('data-custom-attr="yes"')
  })
  
  it('should can be a custom size', () => {
    COMPONENT.setProps({ size: 250 })

    expect(COMPONENT.html()).toContain('height:250px')
    expect(COMPONENT.html()).toContain('width:250px')
  })

  it('should return a input', () => {
    expect(COMPONENT.type()).toEqual('input')
    
    expect(COMPONENT.html()).toContain('type="checkbox"')
  })

})
