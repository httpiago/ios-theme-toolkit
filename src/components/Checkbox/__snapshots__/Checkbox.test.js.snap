// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Checkbox component should render correctly based on last snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Checkbox
    className="aditionalClass"
    color="default"
    size="default"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "className": "checkbox  aditionalClass",
      "style": Object {},
      "type": "checkbox",
    },
    "ref": null,
    "rendered": null,
    "type": "input",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "className": "checkbox  aditionalClass",
        "style": Object {},
        "type": "checkbox",
      },
      "ref": null,
      "rendered": null,
      "type": "input",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
